compiler/src/org.graalvm.compiler.nodes/src/org/graalvm/compiler/nodes/extended/MembarNode.java

```diff
@@ -44,6 +47,12 @@
     public static final NodeClass<MembarNode> TYPE = NodeClass.create(MembarNode.class);
     protected final int barriers;
     protected final LocationIdentity location;
+    // optional load with which this trailing memory barrier may be associated
+    @OptionalInput private ValueNode load;
```

Comment:
I think all these new inputs should have a type like `Association`

Reply:
Hi Gilles,

Sure, if you say so that's great  :-)

I use Memory inputs because I was not sure about what the consequences of adding different InputType links to the graph might be. I reasoned that forward Memory links between things which necessarily had the corresponding memory ordering wouldn't be likely to do any harm. 

So, if these were to be recast as Association links would that have /no/ consequences in any of the other phases that look for Associations? or just /benign/ consequences? I guess fixing this means that I have to add InputType.Association as a valid link type for the various specific load and store classes that might be associated in this way?

One other thing which this current patch does not yet do and might perhaps be needed is ensure that  the changes to DefaultJavaLoweringProvider.java and StandardGraphBuilderPlugins.java that assign the extra input fields are only done when running on AArch64 (by subclassing behaviour of the lowering provider and unsafe plugin, I guess). I don't know whether it really matters that these extra inputs get installed on other archs and stay in place right through to the back end. However, it would probably be clearer to insert them only when running on AArch64. What do you think?

Reply:
I would actually be more concerned about them being memory because they don't really seem to fit the model of memory edges pointing to the node what produced the memory state an node is reading from from.
Association edges on the other end are rather loosely defined so i don't think they could cause any issue.

Reply:
Regarding architectures: yes it would be cleaner to only assign those field in AArch64-specific sub-classes.

In general though i feel like this is more a temporary workaround and we should really aim to keep the volatile read/write as a single node longer to avoid needed those edges altogether. This could be done as an improvement in an other PR though, no need to do it in this PR.